src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -0,0 +1,151 @@
+                        {
+                            Interlocked.Add(ref _totalBufferCount, _numberOfArraysPerBucket);
+                        }
+                        bucket = _buckets[index];
+                    }
+
+                    buffer = bucket.Rent();
+
+                    // If the bucket has an array left and returned it, give it to the caller
+                    if (buffer != null)
+                    {
+                        return buffer;
+                    }
+                }
+
+                // Getting here means that all Buckets from index through _buckets.Length are taken
+                // so we need to allocate a buffer on demand
+                Interlocked.Increment(ref _totalBufferCount);
+                return new T[length];
+            }
+        }
+
+        public override T[] Enlarge(T[] original, int newLength)
+        {
+            if (original == null)
+                throw new ArgumentNullException(nameof(original));
+            // Our buffers have a minimum size of 16, so check for that first
+            if ((newLength > MinimiumArraySize || original.Length > MinimiumArraySize) &&
+                (newLength < original.Length))
+                throw new ArgumentOutOfRangeException(nameof(newLength), SR.Invalid_Enlarge_New_Size);
+
+            // Since we can't realloc, get a bigger buffer, copy the data, clear the old buffer, and return the new one
+            T[] newBuffer = Rent(newLength);
+            Array.Copy(original, newBuffer, original.Length);
+            Return(original, clearArray: true);
+
+            return newBuffer;
+        }
+
+        public override void Return(T[] buffer, bool clearArray = false)
+        {
+            if (buffer == null)
+                throw new ArgumentNullException(nameof(buffer));
+
+            Interlocked.Decrement(ref _rentedBufferCount);
+
+            // Clear the array if the user requests OR if the type is an object. This will ensure there is not internal state leaked accidentally
+            if (clearArray || _isTObjectType) Array.Clear(buffer, 0, buffer.Length);
+
+            // If we can tell that the buffer was allocated, drop it. Otherwise, check if we have space in the pool
+            int bucket = Utilities.SelectBucketIndex(buffer.Length);
+            if (bucket >= _buckets.Length)
+            {
+                Interlocked.Decrement(ref _totalBufferCount);
+            }
+            else if (_buckets[bucket] != null)
+            {
+                // If we don't have room, decrement the count since we're dropping an alloc'd buffer
+                if (_buckets[bucket].Return(buffer) == false)
+                    Interlocked.Decrement(ref _totalBufferCount);
+            }
+#if DEBUG
+            else
+            {
+                System.Diagnostics.Debug.Assert(false); // this shouldn't happen
```

Comment:
I think this should be an exception since this means that the user is trying to return a buffer that they did not get from this pool. Good catch